src/question/order_select/mod.rs

```diff
@@ -0,0 +1,272 @@
+
+struct OrderSelectPrompt<'a, 'c> {
+    prompt: widgets::Prompt<&'a str>,
+    select: widgets::Select<OrderSelect<'c>>,
+    answers: &'a Answers,
+}
+
+impl Prompt for OrderSelectPrompt<'_, '_> {
+    type ValidateErr = widgets::Text<String>;
+    type Output = Vec<ListItem>;
+
+    fn finish(self) -> Self::Output {
+        let OrderSelect {
+            choices,
+            mut order,
+            filter,
+            ..
+        } = self.select.into_inner();
+
+        if let Filter::Sync(filter) = filter {
+            order = filter(order, self.answers);
+        }
+
+        order
+            .into_iter()
+            .filter_map(|i| match &choices.choices[i] {
+                Choice::Choice(text) => Some(ListItem {
+                    index: i,
+                    text: text.text.clone(),
+                }),
+                _ => None,
+            })
+            .collect()
+    }
+
+    fn validate(&mut self) -> Result<ui::Validation, Self::ValidateErr> {
+        if let Validate::Sync(ref mut validate) = self.select.list.validate {
+            validate(&self.select.list.order, self.answers)?;
+        }
+        Ok(ui::Validation::Finish)
+    }
+}
+
+macro_rules! key_swap {
+    ($self: expr, $key: expr, $go_back_if: expr, $go_back_val: expr, $op: tt) => {
+        {
+            if $self.select.list.moving {
+                let mut should_swap = true;
+                let p1 = $self.select.get_at();
+                let p2 = if p1 == $go_back_if {
+                    if $self.select.list.should_loop() {
+                        $go_back_val
+                    } else {
+                        should_swap = false;
+                        p1
+                    }
+                } else {
+                    p1 $op 1
+                };
+
+                if should_swap {
+                    $self.select.list.order.swap(p1, p2);
+                }
+            }
+            $self.select.handle_key($key);
+        }
+    };
+}
+
+impl Widget for OrderSelectPrompt<'_, '_> {
+    fn render<B: Backend>(
+        &mut self,
+        layout: &mut ui::layout::Layout,
+        backend: &mut B,
+    ) -> io::Result<()> {
+        self.prompt.render(layout, backend)?;
+        self.select.render(layout, backend)
+    }
+
+    fn height(&mut self, layout: &mut ui::layout::Layout) -> u16 {
+        self.prompt.height(layout) + self.select.height(layout) - 1
+    }
+
+    fn cursor_pos(&mut self, layout: ui::layout::Layout) -> (u16, u16) {
+        self.select.cursor_pos(layout)
+    }
+
+    fn handle_key(&mut self, key: ui::events::KeyEvent) -> bool {
+        match key.code {
+            ui::events::KeyCode::Up => {
+                key_swap!(self, key, 0, self.select.list.choices.len() - 1, -)
+            }
+            ui::events::KeyCode::Down => {
+                key_swap!(self, key, self.select.list.choices.len() - 1, 0, +)
+            }
```

Comment:
```Rust
if prev_at < new_at {
    self.select.list.order[prev_at..=new_at].rotate_left(1);
} else {
    self.select.list.order[new_at..=prev_at].rotate_right(1);
}
```

Can be replaced with

```Rust
self.select.list.order.swap(prev_at, new_at);
```

Reply:
There is a difference, take the following list:

```
Item 1
Item 2
Item 3
Item 4
Item 5
```

Suppose, `prev_at = Item 5`, and `new_at = Item 2`. Swapping `prev_at` and `new_at` will lead to:
```
Item 1
Item 5
Item 3
Item 4
Item 2
```

Using the rotations it would instead be:
```
Item 1
Item 5
Item 2
Item 3
Item 4
```

Notice the difference in position of Item 2. I feel that the output of the rotation is more natural as it is like taking Item 5 out and inserting it at the position where Item 2 was.

Using `swap` would work if `prev_at` and `new_at` are only a single number apart. However, the use of `PageUp`, `PageDown`, `Home` and `End` can cause a larger shift.